In [1]:
import pandas as pd
import numpy as np

# <u>Melting (Unpivoting) & Pivoting dataframes in Pandas</u>

<br>

## Melting
- Converting a dataframe from columnar to row format, because the original dataframe was not workable. Due to fact that it has more columns than rows, it does not look representable.

In [2]:
weather  = pd.read_csv("../datasets/weather_dataset/weather.csv")
weather1 = pd.read_csv("../datasets/weather_dataset/weather1.csv")
weather2 = pd.read_csv("../datasets/weather_dataset/weather2.csv")
weather3 = pd.read_csv("../datasets/weather_dataset/weather3.csv")
weather4 = pd.read_csv("../datasets/weather_dataset/weather4.csv")

In [3]:
weather.shape

(22, 34)

In [10]:
weather.head()

year  month element       day1       day2       day3       day4       day5  \
0  2018      1     max  17.573016  19.796815  22.412495  17.813163  20.165825   
1  2018      1     min  22.725760  21.007865  17.730792  18.045290  20.766734   
2  2018      2     max  19.015120  19.261805  17.510713  21.080425  17.915749   
3  2018      2     min  18.653843  22.818600  21.842673  21.958159  22.523078   
4  2018      3     max  20.741115  19.704016  17.039811  20.703908  22.714125   

        day6       day7  ...      day22      day23      day24      day25  \
0  17.060539  22.736134  ...  22.629226  19.013674  17.831558  17.018316   
1  18.656651  22.607481  ...  21.163491  17.371653  17.194530  18.964190   
2  19.082145  18.056023  ...  21.254968  20.346324  20.036019  22.926773   
3  18.535469  19.636158  ...  19.711974  20.586232  18.780840  20.065633   
4  17.205000  19.079503  ...  19.528963  19.563689  22.658997  22.446357   

       day26      day27      day28      day29      day30      day31  
0  21.837387  19.700255  19.084056  18.534690  20.787275  21.117423  
1  17.892385  17.646139  19.844125  22.729859  19.768917  20.913443  
2  17.533172  20.308927  20.438899  17.005806  18.346534  20.878165  
3  18.293548  20.823661  18.056112  22.748292  22.470172  18.602007  
4  17.911240  22.591993  17.406800  19.118712  22.767154  17.119452  

[5 rows x 34 columns]

<br>It's clear that `weather` dataset looks to be in __columnar (or, wide) format__. Obviously, it would become difficult to work with it.

So, let's __change it from column to row (or, long) format via Melting__.

(Actually, transpose or __.T__ won't be of better use here, because still there will remain the problem of relatively greater number of rows that will obviously make the dataframe difficult to work with. Hence, the usecase of Melting arises.)

In [25]:
weather_melt = weather.melt(id_vars=["year","month","element"],
                            var_name="day",
                            value_name="temperature")

weather_melt

year  month element    day  temperature
0    2018      1     max   day1    17.573016
1    2018      1     min   day1    22.725760
2    2018      2     max   day1    19.015120
3    2018      2     min   day1    18.653843
4    2018      3     max   day1    20.741115
..    ...    ...     ...    ...          ...
677  2018     10     min  day31    21.691537
678  2018     11     max  day31    20.750438
679  2018     11     min  day31    18.939767
680  2018     12     max  day31    19.648924
681  2018     12     min  day31    18.775539

[682 rows x 5 columns]

Important points to note here :

- shape of melted dataframe is different from the original one.


- How (682, 5) is the new shape ?<br>
In fact, we melted the dataframe along all those columns except the 3 mentioned in `id_vars`. So, melting happened around `34-3 = 31` columns; which became the rows. So, `31 * 22` = __682 rows__.<br>
Similarly, 3 columns mentioned in `id_vars` + 1 new column created out of `var_name` + 1 new column created out of `value_name` = __5 columns__.


- Due to Melting, new dataframe looks to be more workable now.


- `var_name` : name to use for `Variable` column i.e. "day" column


- `value_name` : name to use for `Value` column i.e. "temperature" column.  `Value` column is the column that uses values of `Variable` column

<br>

Let's take example of `weather4` dataset.

In [22]:
weather4.shape

(7, 4)

In [11]:
weather4

day  chicago  chennai  berlin
0     Monday       32       75      41
1    Tuesday       30       77      43
2  Wednesday       28       75      45
3   Thursday       22       82      38
4     Friday       30       83      30
5   Saturday       20       81      45
6     Sunday       25       77      47

In [24]:
weather4.melt(id_vars=["day"],
              var_name="City",
              value_name="Temperature"
             ).rename(columns={"day":"Day"})

## .shape => (21,3)

Day     City  Temperature
0      Monday  chicago           32
1     Tuesday  chicago           30
2   Wednesday  chicago           28
3    Thursday  chicago           22
4      Friday  chicago           30
5    Saturday  chicago           20
6      Sunday  chicago           25
7      Monday  chennai           75
8     Tuesday  chennai           77
9   Wednesday  chennai           75
10   Thursday  chennai           82
11     Friday  chennai           83
12   Saturday  chennai           81
13     Sunday  chennai           77
14     Monday   berlin           41
15    Tuesday   berlin           43
16  Wednesday   berlin           45
17   Thursday   berlin           38
18     Friday   berlin           30
19   Saturday   berlin           45
20     Sunday   berlin           47

## Pivoting

In [36]:
weather_melt.head()

year  month element   day  temperature
0  2018      1     max  day1    17.573016
1  2018      1     min  day1    22.725760
2  2018      2     max  day1    19.015120
3  2018      2     min  day1    18.653843
4  2018      3     max  day1    20.741115

In `weather_melt` we observe that, `element` column has __repeated values__ - "min" & "max", as its values.

These values can be entirely be converted to 2 different column - `min` & `max`. And, values of these 2 new columns will be values of `temperature` column.

In layman terms, we are compressing the `element` column.

In [78]:
weather_tidy = weather_melt.pivot(index=["year","month","day"],      ## columns of original dataframe that will remain as it is.
                                  columns=["element"],               ## column around which pivotting happens. This gets created as new column.
                                  values="temperature")              ## values of the new column


weather_tidy

element                 max        min
year month day                        
2018 1     day1   17.573016  22.725760
           day10  19.067288  19.931129
           day11  19.361002  22.598325
           day12  20.982134  17.715137
           day13  21.668005  17.940334
...                     ...        ...
     12    day5   21.375349  20.865535
           day6   17.992885  20.310116
           day7   19.683359  20.531823
           day8   20.477046  19.310346
           day9   20.210640  22.820992

[341 rows x 2 columns]

In [79]:
weather_tidy.columns

Index(['max', 'min'], dtype='object', name='element')

In [80]:
weather_tidy.reset_index(inplace=True)   ## reset index so that all levels are eliminated

weather_tidy

element  year  month    day        max        min
0        2018      1   day1  17.573016  22.725760
1        2018      1  day10  19.067288  19.931129
2        2018      1  day11  19.361002  22.598325
3        2018      1  day12  20.982134  17.715137
4        2018      1  day13  21.668005  17.940334
..        ...    ...    ...        ...        ...
336      2018     12   day5  21.375349  20.865535
337      2018     12   day6  17.992885  20.310116
338      2018     12   day7  19.683359  20.531823
339      2018     12   day8  20.477046  19.310346
340      2018     12   day9  20.210640  22.820992

[341 rows x 5 columns]

Important to note here :
- original dataframe had 682 rows. After pivoting (around 1 column), the number of rows got halved i.e. became 341, because originally the `element` column had 2 unique values - min & max.


- `index` : _list_. Mention the columns that you do not want to change.


- `columns` : _list_. Mention the columns around which you want to do pivot.


- `values` : _list_. Mention column names whose values would populate in the new columns.

In [50]:
weather1

date      city  temperature  humidity
0  5/1/2017  new york           65        56
1  5/2/2017  new york           66        58
2  5/3/2017  new york           68        60
3  5/1/2017    mumbai           75        80
4  5/2/2017    mumbai           78        83
5  5/3/2017    mumbai           82        85
6  5/1/2017   beijing           80        26
7  5/2/2017   beijing           77        30
8  5/3/2017   beijing           79        35

In [54]:
weather1_pivot = weather1.pivot(index="city",
                                columns="date",
                                values=["temperature","humidity"])

weather1_pivot

temperature                   humidity                  
date        5/1/2017 5/2/2017 5/3/2017 5/1/2017 5/2/2017 5/3/2017
city                                                             
beijing           80       77       79       26       30       35
mumbai            75       78       82       80       83       85
new york          65       66       68       56       58       60

In [76]:
weather1_pivot.columns

MultiIndex([('temperature', '5/1/2017'),
            ('temperature', '5/2/2017'),
            ('temperature', '5/3/2017'),
            (   'humidity', '5/1/2017'),
            (   'humidity', '5/2/2017'),
            (   'humidity', '5/3/2017')],
           names=[None, 'date'])

<br>If we want only humidity data :

In [69]:
weather1.pivot(index="city",
               columns="date",
               values=["humidity"])

humidity                  
date     5/1/2017 5/2/2017 5/3/2017
city                               
beijing        26       30       35
mumbai         80       83       85
new york       56       58       60

<br>Pivot around `city` column.

In [70]:
weather1.pivot(index="date",
               columns="city")

temperature                 humidity                
city         beijing mumbai new york  beijing mumbai new york
date                                                         
5/1/2017          80     75       65       26     80       56
5/2/2017          77     78       66       30     83       58
5/3/2017          79     82       68       35     85       60

## Pivot table

- `.pivot_table()`

In [3]:
weather2

## this dataset shows min and max temperatures for each day in 3 cities.

date      city  temperature  humidity
0  5/1/2017  new york           65        56
1  5/1/2017  new york           61        54
2  5/2/2017  new york           70        60
3  5/2/2017  new york           72        62
4  5/1/2017    mumbai           75        80
5  5/1/2017    mumbai           78        83
6  5/2/2017    mumbai           82        85
7  5/2/2017    mumbai           80        26

<br>
Let's say you want to convert this dataset into a more readable format :

- name of cities should come in the "rows".
- in the column, we would want to have mean (average) "values" of humidity and temperatures.

In [7]:
weather2_pivotTable = weather2.pivot_table(index=["city"],
                                           columns=["date"],
                                           values=["humidity", "temperature"],      ## not necessary to declare
                                           aggfunc="mean",
                                           margins=True)

weather2_pivotTable

humidity                 temperature                 
date     5/1/2017 5/2/2017    All    5/1/2017 5/2/2017     All
city                                                          
mumbai      81.50    55.50  68.50       76.50     81.0  78.750
new york    55.00    61.00  58.00       63.00     71.0  67.000
All         68.25    58.25  63.25       69.75     76.0  72.875

In [86]:
weather2.pivot_table(index=["city"],
                     columns=["date"],
                     aggfunc="mean")

humidity          temperature         
date     5/1/2017 5/2/2017    5/1/2017 5/2/2017
city                                           
mumbai       81.5     55.5        76.5     81.0
new york     55.0     61.0        63.0     71.0

Important points to note here :

- `margins=` => if `True`, then cummulative aggregates (i.e. `All` named column) are also displayed but not for `False`.

# <u>Convert Numerical data into Categorical data</u>

## 1. Binning

- `pd.cut()`

In [98]:
weather_tidy.head()

element  year  month    day        max        min
0        2018      1   day1  17.573016  22.725760
1        2018      1  day10  19.067288  19.931129
2        2018      1  day11  19.361002  22.598325
3        2018      1  day12  20.982134  17.715137
4        2018      1  day13  21.668005  17.940334

In [99]:
weather_tidy["max"].max()

22.9403089333132

In [100]:
weather_tidy["min"].min()

17.0538502637388

Lets convert continuous numerical values of `max` column into categorical values :

In [102]:
bins   = [16, 18, 20, 22, 24]
labels = ["16-18", "18-20", "20-22", "22-24"]

weather_tidy["max_categorical"] = pd.cut(weather_tidy["max"],
                                        bins=bins,
                                        labels=labels)

weather_tidy.head(20)

element  year  month    day        max        min max_categorical
0        2018      1   day1  17.573016  22.725760           16-18
1        2018      1  day10  19.067288  19.931129           18-20
2        2018      1  day11  19.361002  22.598325           18-20
3        2018      1  day12  20.982134  17.715137           20-22
4        2018      1  day13  21.668005  17.940334           20-22
5        2018      1  day14  19.441664  19.853651           18-20
6        2018      1  day15  17.609544  18.662379           16-18
7        2018      1  day16  20.245582  17.215839           20-22
8        2018      1  day17  17.386774  21.662470           16-18
9        2018      1  day18  22.549572  17.202508           22-24
10       2018      1  day19  19.898544  19.697496           18-20
11       2018      1   day2  19.796815  21.007865           18-20
12       2018      1  day20  20.836153  22.182156           20-22
13       2018      1  day21  18.427361  21.532442           18-20
14       2018      1  day22  22.629226  21.163491           22-24
15       2018      1  day23  19.013674  17.371653           18-20
16       2018      1  day24  17.831558  17.194530           16-18
17       2018      1  day25  17.018316  18.964190           16-18
18       2018      1  day26  21.837387  17.892385           20-22
19       2018      1  day27  19.700255  17.646139           18-20

- the intervals defined in `labels` are - `[inclusive_number, exlusive_number)` i.e. [18,20) , [20,22)

__Parameters :__
- column to cut
- bins
- labels to denote each bin

In [103]:
bins   = [16, 18, 20, 22, 24]
labels = ["low", "medium", "high", "extreme"]

weather_tidy["max_categorical_II"] = pd.cut(weather_tidy["max"],
                                            bins=bins,
                                            labels=labels)

weather_tidy.head()

element  year  month    day        max        min max_categorical  \
0        2018      1   day1  17.573016  22.725760           16-18   
1        2018      1  day10  19.067288  19.931129           18-20   
2        2018      1  day11  19.361002  22.598325           18-20   
3        2018      1  day12  20.982134  17.715137           20-22   
4        2018      1  day13  21.668005  17.940334           20-22   

element max_categorical_II  
0                      low  
1                   medium  
2                   medium  
3                     high  
4                     high